### ISSUE TITLE
Homebrew sonobuoy version lacks 'v' prefix

### ISSUE THREAD

**Issue by johnSchnake** [heptio/sonobuoy#715]
**What steps did you take and what happened:**
 - Ensure you dont have sonobuoy locally (to be sure to use the one from brew later)
 - brew install sonobuoy
 - `sonobuoy version` (this will lack the `v` prefix)
 - `sonobuoy gen|grep 14` (this will have the image names for sonobuoy being `...:0.14.2` which also lacks the `v` prefix and those images do not exist)

**What did you expect to happen:**
The sonobuoy from brew to be an actual release, report the same version info, and choose proper images.

**Anything else you would like to add:**
Originally reported here: https://github.com/heptio/sonobuoy/issues/556#issuecomment-491567885

I also have no idea how sonobuoy gets into homebrew; I'm not involved with this process and I dont know if it is some community effort, a custom build, pulls from master...etc. Need some insight on that. I confirmed that master is just fine and that the release is also fine.

**Comment by johnSchnake** [comment]
https://github.com/Homebrew/homebrew-core/blob/master/Formula/sonobuoy.rb

**Comment by joaovitor** [comment]
@johnSchnake thanks for opening this issue.

Even using sonobuoy from brew I tried to override the image version using the config file and it made no difference. This may be other issue.

**Comment by johnSchnake** [comment]
Thanks for pushing the issue; a few separate issues here:

> I tried to override the image version using the config file and it made no difference

I'm guessing what you did was to set the image via the config to `gcr.io/heptio-images/sonobuoy:v0.14.2` ?

It looks like this _was_ a genuine bug in that sonobuoy was not properly handling that field from the config when assigning the aggregator pod. That looks like it was actually just fixed here: https://github.com/heptio/sonobuoy/pull/704 The problem was there were some redundant fields and they were not properly being handled/sync'd.

However, there is another problem that I found when trying to analyze this (because I misread it at first and thought it was the kube-conformance image that had the problem not the sonobuoy image): created a new issue here: https://github.com/heptio/sonobuoy/issues/716

So:
 - You are welcome to use homebrew but at this time I don't actively recommend it because it has its own build issue for not putting the `v`, though I wish it would just use the pre-built from our github releases anyways. I'm hoping that the community will help with this since I know little of maintaining a homebrew forumla and the possible options/config you can give it
 - if you use that version you'll need to use a flow like `sonobuoy gen > out.yaml`, edit the yaml to set the aggregator pod image, `kubectl apply -f out.yaml` since the templating logic wasn't properly handling the sonobuoy image if overridden for the aggregator pod (the right image gets used for the plugins, but not for the main pod)

To be clear, the other bugs mentioned here still exist in v1.14.2 but wouldn't stop normal runs because the default image for the aggregator pod would have been valid.

**Comment by sc250024** [comment]
> I also don't understand a lot of the other info there; they have different shas it seems for each OS version and then also has some `sonobuoyresults` defined which is a link to random testdata from source.

@johnSchnake This was because when I was writing the initial formula, one of the maintainers insisted that I perform some unit tests on the binary to ensure that it works. Pulling down the dummy report, and analyzing the output seemed like a good way to do that. I'm open for better ways if you know how.

**Comment by sc250024** [comment]
@johnSchnake Also, since the PR (https://github.com/Homebrew/homebrew-core/pull/40977) was closed already, let's just continue the discussion here. If you add the `Makefile` targets, I'll update the Homebrew formula.

P.S. it would be nice to have a generic "Brew" target that works for both macOS (Homebrew) and Linux (Linuxbrew) repositories.

**Comment by zubron** [comment]
> P.S. it would be nice to have a generic "Brew" target that works for both macOS (Homebrew) and Linux (Linuxbrew) repositories.

@sc250024 That should be possible since we'll be using a native go build. The formula will probably just need some changes to specify the `GOOS`. I'm not familiar with the differences between the macOS and Linux versions of Homebrew though. I'll be looking at that more today!

With regard to the testing step in the formula, how much functionality of the binary needs to be shown? Also, when is the testing step performed? I'm not sure yet what the best way to test it would be, but if we're going to use some data from the source repo, it seems like we should be using a fixed reference in the repository rather than a branch name. Otherwise it looks like there is a risk the file might be modified and the SHA check would fail? https://github.com/sc250024/homebrew-core/blob/dcdd384dc5fdcaf74e278576e07d0c7b5afbca81/Formula/sonobuoy.rb#L17-L18

Sorry for asking so many questions. Thanks for your help with this! :smile:

**Comment by sc250024** [comment]
@zubron Actually I stand corrected; looks like it was merged now into one project (https://github.com/Linuxbrew/brew). And, from what I can see, there's probably some mirroring going on to the `linuxbrew-core` repository (https://github.com/Homebrew/linuxbrew-core) from the main `Homebrew-core` repo.

> With regard to the testing step in the formula, how much functionality of the binary needs to be shown?

It depends; I've seen some packages that only output a version, but I think those were early on. The maintainers want to test at least _some_ basic functionality to make sure it works.

If you look at the following lines (https://github.com/Homebrew/homebrew-core/blob/master/Formula/sonobuoy.rb#L34-L42) I do the following:

* Checks that running just `sonobuoy 2>&1` outputs the help text; I then check if a string is found.
* Checks that the version string is found somewhere in the `sonobuoy version 2>&1` command.
* Checks that the string `name: heptio-sonobuoy` is found within the YAML generated from `sonobuoy gen --kube-conformance-image-version=v1.12 2>&1` 
* Finally, checks that the string `all tests` is found from running the command `sonobuoy e2e --show=all " + resource("sonobuoyresults").cached_download + " 2>&1` against these test results in the repository (https://raw.githubusercontent.com/heptio/sonobuoy/master/pkg/client/results/testdata/results-0.10.tar.gz).

The `build_sonobuoy` target (https://github.com/heptio/sonobuoy/blob/master/Makefile#L129-L130) already has the right command, except for the `$(DOCKER)` part in the beginning.

**Comment by zubron** [comment]
@sc250024 I've just merged #757 which adds a new Makefile target (`native`). We opted for this rather than something Homebrew specific as it will be useful for the team and potentially other contributors outside of Homebrew related installs.

We'll be releasing a new version of Sonobuoy hopefully within the next few days (waiting on Kubernetes 1.15) so once that happens we'll be able to take advantage of this new target in the formula.
